### PR TITLE
A separate occupancy tracker for MR 

### PR DESCRIPTION
**Pull request by khlifi411**
Added separate occupancy tracking and refactored metrics registrations

### CODE REVIEW

**Comment by khlifi411** [comment]
PR was closed as the merge target was changed after release to the develop branch

**Comment by khlifi411** [comment]
PR was reopened as the merge target was changed back to the main branch

File: pkg/reconciler/service/occupancytracker.go [hunk 1]
```diff
@@ -14,41 +14,46 @@ import (
 	"time"
 )
 
-const occupancyURLTemplate = "%s://%s:%s/v1/occupancy/%s"
+const (
+	occupancyURLTemplate   = "%s://%s:%s/v1/occupancy/%s"
```
Comment (tobiscr) [inline pkg/reconciler/service/occupancytracker.go:18]: Are we using it somewhere?

File: pkg/scheduler/service/occupancytracker.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	defaultOccupancyCleanUpInterval  = 5 * time.Minute //default: 5 mins
```
Comment (jeremyharisch) [inline pkg/scheduler/service/occupancytracker.go:19]:
```suggestion
	defaultOccupancyCleanUpInterval  = 5 * time.Minute
```
Unneeded comment

File: pkg/metrics/workerpool_occupancy.go [hunk 3]
```diff
@@ -54,3 +61,13 @@ func (c *WorkerPoolOccupancyCollector) Collect(ch chan<- prometheus.Metric) {
 	}
 	c.workerPoolOccupancyGaugeVec.Collect(ch)
 }
+
+func buildLabelValuesMap(cfg *config.Config) map[string]string {
+	labelValuesMap := make(map[string]string, len(cfg.Scheduler.Reconcilers)+1)
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/metrics/workerpool_occupancy.go:66]: Here, we need to add the mothership label to the map at the end of the method. As it is not included in the original list of reconcilers, we need to increase the size by one.

File: pkg/metrics/workerpool_occupancy.go [hunk 4]
```diff
@@ -1,34 +1,41 @@
...
 type WorkerPoolOccupancyCollector struct {
 	occupancyRepository         occupancy.Repository
 	logger                      *zap.SugaredLogger
-	componentList               []string
+	labelValuesMap              map[string]string
 	workerPoolOccupancyGaugeVec *prometheus.GaugeVec
 }
 
-func NewWorkerPoolOccupancyCollector(occupancyRepository occupancy.Repository, reconcilerList []string, logger *zap.SugaredLogger) *WorkerPoolOccupancyCollector {
+func NewWorkerPoolOccupancyCollector(occupancyRepository occupancy.Repository, cfg *config.Config, logger *zap.SugaredLogger) *WorkerPoolOccupancyCollector {
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/metrics/workerpool_occupancy.go:25]: Thanks! That's a good idea. Applied it 👍

File: pkg/metrics/workerpool_occupancy.go [hunk 5]
```diff
@@ -54,3 +61,13 @@ func (c *WorkerPoolOccupancyCollector) Collect(ch chan<- prometheus.Metric) {
 	}
 	c.workerPoolOccupancyGaugeVec.Collect(ch)
 }
+
+func buildLabelValuesMap(cfg *config.Config) map[string]string {
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/metrics/workerpool_occupancy.go:65]: Thanks! Applied it 👍

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -3,25 +3,28 @@ package metrics
...
+	reconciliationWaitingCollector := NewReconciliationWaitingCollector(inventory, logger)
+	reconciliationNotReadyCollector := NewReconciliationNotReadyCollector(inventory, logger)
+	prometheus.MustRegister(reconciliationWaitingCollector, reconciliationNotReadyCollector)
+}
+
+func RegisterOccupancy(occupancyRepo occupancy.Repository, cfg *config.Config, logger *zap.SugaredLogger) {
 	if features.WorkerpoolOccupancyTrackingEnabled() {
-		workerPoolOccupancyCollector := NewWorkerPoolOccupancyCollector(occupancyRepo, reconcilerList, logger)
-		collectors = append(collectors, workerPoolOccupancyCollector)
+		prometheus.MustRegister(NewWorkerPoolOccupancyCollector(occupancyRepo, cfg, logger))
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/metrics/metrics.go:28]: Thanks! Applied it 👍

File: pkg/scheduler/occupancy/occupancy.go [hunk 7]
```diff
@@ -9,9 +9,11 @@ type Repository interface {
 	CreateWorkerPoolOccupancy(poolID, component string, runningWorkers, poolSize int) (*model.WorkerPoolOccupancyEntity, error)
 	FindWorkerPoolOccupancyByID(poolID string) (*model.WorkerPoolOccupancyEntity, error)
 	GetComponentList() ([]string, error)
+	GetComponentIDs() ([]string, error)
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/scheduler/occupancy/occupancy.go:12]: Thanks! That's a good idea. Renamed it 👍

File: pkg/scheduler/service/occupancytracker.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+				found = true
+				break
+			}
+		}
+		if !found {
+			idsOfOrphanComponents = append(idsOfOrphanComponents, componentID)
+		}
+	}
+	if len(idsOfOrphanComponents) == 0 {
+		t.logger.Warnf("occupancy tracker found 0 orphan occupancies: nothing to clean")
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/scheduler/service/occupancytracker.go:170]: Thanks! That's a good idea. Applied it 👍

File: pkg/scheduler/occupancy/persistent.go [hunk 9]
```diff
@@ -12,6 +12,52 @@ type PersistentOccupancyRepository struct {
 	*repository.Repository
 }
 
+func (r *PersistentOccupancyRepository) RemoveWorkerPoolOccupancies(poolIDs []string) (int, error) {
+	dbOps := func(tx *db.TxConnection) (interface{}, error) {
+		if len(poolIDs) == 0 {
+			r.Logger.Warnf("OccupancyRepo received empty list of ids: nothing to remove")
```
[the remark this answers is not in the document]
Comment (khlifi411) [reply pkg/scheduler/occupancy/persistent.go:18]: Thanks! That's a good idea. Applied it 👍

File: pkg/scheduler/service/occupancytracker.go (hunk 2 above)
Comment (khlifi411) [reply pkg/scheduler/service/occupancytracker.go:19]: Thanks! Removed it 👍

**Review by jeremyharisch — APPROVED** [review]

**Pull request merged**